Step3 - 질문 삭제하기 리팩터링

mnonm opened the pull request.

Commit:
docs(README.md): 요구 사항 작성

Commit:
build(build.gradle): h2 버전 수정

- h2 설정 문제로 이전 버전 적용 1.4.2 -> 1.4.199

Commit:
build(application.properties): h2 설정 추가

Commit:
feat(Answer): Answer 엔티티 및 테스트 작성

- BaseTimeEntity 추가

Commit:
fix(DeleteHistory, Question, user): 테스트 실행을 위한 임시 수정

Commit:
docs(README.md): 작업 순서 변경

Commit:
fix(BaseTimeEntity): 수정 시간에 not null 제거

Commit:
feat(Question): Question 엔티티 및 테스트 작성

Commit:
feat(DeleteHistory): DeleteHistory 엔티티 및 테스트 작성

Commit:
test(AnswerTest, Questiontest): id 비교 테스트 코드 제거

Commit:
feat(User): User 엔티티 및 테스트 코드 작성

Commit:
docs(README.md): 요구 사항 수정

Commit:
docs(README.md): step2 요구 사항

Commit:
feat(Answer): answer - question 매핑

mnonm committed:
feat(Answer): answer - user 매핑

mnonm committed:
docs(README.md): 작업 순서 변경

mnonm committed:
feat(question, delete_history) - question - user, delete_history - user 매핑

- 기존 id필드를 User 엔티티로 변경에 따른 수정

mnonm committed:
refactor: 테스트 코드 클래스명 변경

- 도메인 객체 테스트와 구분하기 위해 RepositoryTest로 수정
- 더미 데이터 한 곳에서 관리(TestDummy)

mnonm committed:
fix(AnswerRepositoryTest): 테스트 실패 수정

- Dummy User 데이터에 id 제거
- Dummy User에 id가 지정되어 있어 저장시 AuditingEntityListener에서 update로 처리해서 createdDateTime이 생성되지 않음

mnonm committed:
Merge branch 'mnonm' into step2

mnonm committed:
docs(README.md): 기능 요구 사항

mnonm committed:
fix(Question): 질문의 작성자인지 확인하는 로직 수정

- id끼리 비교하도록 수정
- QnaServiceTest: 유저 더미데이터에 아이디가 제외됐으므로 아이디 추가해서 생성하는 헬퍼 메소드 추가

mnonm committed:
test(실습): 지하철 노선

mnonm committed:
refactor(QnaService, Question): 질문 작성자만 삭제 가능

- 질문 삭제 로직 Question으로 이동

mnonm committed:
refactor(Answers, QnaService) : 질문자와 답변 글의 모든 답변자가 같은 경우 삭제 가능

- Answers 일급 컬렉션 추가
- Answers 전체 삭제 함수 추가
- Answer 삭제 함수 추가

Commit:
fix(Answer, DeleteHistory, Question): 연관 관계 매핑 수정

- Answer - User 다대일 변경
- DeleteHistory - User 다대일 변경
- Question - User 다대일 변경

Commit:
fix(Answer, Question): 연관 관계 매핑 수정

- 다대일 양방향 관계로 수정
- 편의메소드 추가

mnonm committed:
Merge remote-tracking branch 'origin/step2' into step3

mnonm committed:
refactor(DeleteHistory) : 생성자 수정

- createDate 기본값이 있으므로 생성자에서 제거

mnonm committed:
fix(data.sql): 실습 데이터 파일명 수정

- 파일명이 data.sql인 경우 테스트 실행시 table not found 에러 발생

mnonm committed:
refactor: 질문과 답변 삭제 이력에 대한 정보 저장

- 삭제 이력을 반환할 수 있는 인터페이스 추가, SavingDeleteHistory.java
- Answer, Question에서 SavingDeleteHistory 구현
- QnaService.deleteQuestion 함수 리팩토링

mnonm committed:
refactor(Question, Answers): 질문의 답변 일급콜렉션으로 수정

mnonm committed:
refactor(QuestionRepository): 질문 저장, 조회시 답변까지 포함하도록 수정

- QuestionRepository.findByIdAndDeletedFalseWithAnswers 추가

mnonm committed:
fix(Answers): Answers cascade all로 수정

- persist 외의 다른 옵션도 적용되게끔 all 수정
- 테스트 격리 위해 TestDummy 데이터 사용 부분 제거

mnonm committed:
refactor(QnaService): deleteQuestion 함수 리팩토링

- DeleteHistories 일급컬렉션 추가

mnonm committed:
refactor(Answer): Contents 일급객체 추가

mnonm committed:
refactor(Question): Contents 일급객체로 변경

mnonm committed:
refactor(Question): Title 일급객체로 변경

- Title 일급 객체 추가

mnonm committed:
refactor(Answer, User): 미사용 코드 정리

mnonm committed:
fix(Question): 이미 삭제된 답변이 포함된 경우, 질문 삭제시 중복으로 답변의 삭제 기록이 남는 문제 수정

- 지워지지 않은 답변만 가지고 기록을 생성하도록 수정
- 테스트 코드 추가

mnonm committed:
refactor(User): String 변수 래핑

- UserId, Password, Name, Email 클래스 추가

mnonm committed:
docs(README.md): 오타 수정

mnonm committed:
Merge branch 'mnonm' into step3

mnonm committed:
fix: 머지 에러 수정

5minho commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -49,15 +57,29 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-    public void toQuestion(Question question) {
-        this.question = question;
+    public DeleteHistory delete(User loginUser) throws CannotDeleteException {
+        if (!isOwner(loginUser)) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
+        }
+        setDeleted(true);
+
+        return toDeleteHistory();
+    }
+
+    @Override
+    public DeleteHistory toDeleteHistory() {
+        if (deleted) {
+            return new DeleteHistory(ContentType.ANSWER, id, writer);
+        }
+
+        throw new IllegalStateException("삭제시에만 기록을 남길 수 있습니다.");
     }
```
delete 메서드에서 DeleteHistory 를 반환하는것 좋네요.
근데 굳이 `toDeleteHistory` 를 public 으로 노출해서 외부에서 아무나 DeleteHistory 객체를 획득가능하게 하는게 좋은건지는 의문이에요.
현재 `SavingDeleteHistory` 인터페이스가 정말 필요한지 고민해보시고 toDeleteHistory 를 외부에 열어 놓는 것과 닫아 놓는 것 둘 중 어느 것이 사이드 이펙트가 생길 확률이 적을까 고민해보셨으면 좋겠습니다.
추가로 현재 setDeleted 메서드도 public 으로 외부에 공개된 상태인데 private 으로 감추는게 좋지 않을까요?

mnonm replied:
제가 처음 의도한 바는 해당 인터페이스를 구현하면 이 클래스는 soft delete 해야 한다를 명시하고 삭제시에 삭제 기록을 남겨야 한다는 것을 강제화하고 싶었습니다. 또 삭제 기록을 남기는 것과 질문, 답변의 삭제 기능이 결합력이 높아져도 되나 하는 생각에 인터페이스로 나누고 삭제 기록을 남기는 것과 삭제의 테스트를 분리할 수 있었습니다.

하지만 삭제 기록 생성 함수가 노출되는 점에서는 사이드 이팩트가 우려되는거 같습니다. 외부에 노출될 수 있는 기능만 인터페이스를 작성하는게 맞는거 같습니다!

해당 인터페이스 제거했고, 삭제 기록 생성 함수와 setDeleted 메서드 모두 private으로 변경했습니다

5minho commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,58 @@
+
+@Embeddable
+public class Answers {
+    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL)
```
CascadeType.All 을 해주셨는데요. 이유가 무엇인가요?
동료개발자가 실수로 Question 을 soft delete 가 아닌 hard delete 하면 어떤일이 벌어질까요?

mnonm replied:
Answer의 생명주기는 Question에 속해있다고 생각했습니다.
Question이 삭제되면 Answer가 의미가 있나라는 생각에서 CascadeType.All을 했습니다.

하지만 말씀주신대로 soft delete를 하기 때문에 PERSIST와 MERGE만 사용하는게 더 요구사항에 어울리는 설정인 거 같습니다.

5minho replied:
만약 Answer 가 Question 과 연관관계가 있었다면 CascadeType 을 All 로 해도 좋았을거 같아요.
하지만 지금 Answer 는 User 와도 연관관계를 맺고 있습니다. 이럴 경우에는 CascadeType.All 은 위험하다고 생각해요. 
내가 쓴 답변들이 갑자기 사라질 수도 있기 때문이에요.

5minho commented on src/main/java/qna/domain/Contents.java:
```diff
@@ -0,0 +1,46 @@
+import javax.persistence.Embeddable;
+import javax.persistence.Lob;
+
+@Embeddable
```
값객체를 사용해보신것 좋네요 👍

5minho commented on src/main/java/qna/domain/Password.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private Password(String value) {
+        this.value = value;
```
최소한의 유효성 체크 예를들면 빈 문자열 체크 같은것이 있으면 값 객체로 래핑한 의미가 더 살거 같아요

mnonm replied:
공백 또는 null 일 경우 예외 처리 추가했습니다!

5minho commented on src/main/java/qna/domain/Question.java:
```diff
@@ -26,20 +35,61 @@
+        this.contents = Contents.from(contents);
+    }
+
+    public void addAnswer(Answer answer) {
+        this.answers.add(answer);
+        if (answer.getQuestion() != this) {
+            answer.setQuestion(this);
+        }
+    }
```
addAnswer 같이 내부의 변경을 일으키는 메서드는 public 으로 노출시키는 것보다 최대한 외부에서 보지 못하도록 감추는게 좋지 않을까요? answer.setQuestion 도 마찬가지구요.

mnonm replied:
default로 변경하니까 QnaServiceTest에서 addAnswer를 사용하고 있어서, 에러가 발생했습니다.
패키지를 이동하는건 아닌거 같아서 생성 시점에 편의 메소드를 호출하게 변경했는데,
의도하신 방향이 맞을까요?

5minho replied:
네 맞습니다!
만약 지금 답변 추가 같은 요구사항이 있었다면 public 으로 열어놔야 하겠지만 그런 요구사항은 없고, 
중요한 도메인 객체 내부를 변경하는 메서드이기 때문에 최대한 내부에 숨겨서 사이드이펙트를 방지하는게 좋을거 같아서
위와 같은 피드백을 드렸습니다

5minho commented on src/main/java/qna/domain/QuestionRepository.java:
```diff
@@ -1,12 +1,14 @@
 
     Optional<Question> findByIdAndDeletedFalse(Long id);
+
+    @Query("SELECT q FROM Question q JOIN FETCH q.answers.values a "
```
fetch join 을 사용해보셨군요 👍

5minho commented on src/main/java/qna/domain/SavingDeleteHistory.java:
```diff
@@ -0,0 +1,5 @@
+package qna.domain;
+
+public interface SavingDeleteHistory {
```
`SavingDeleteHistory` 인터페이스를 만들어보신 시도는 좋지만, 의도가 궁금해요.
현재 작성해주신 코드를 보면 인터페이스를 구현만 하고 실제 사용은 하지 않아서요

mnonm replied:
위에 있던 답변이랑 비슷한 내용일거 같아 여기선  생략하겠습니다

5minho commented on src/main/java/subway/Application.java:
```diff
@@ -0,0 +1,14 @@
+package subway;
```
수업때 하신 내용인가요? ㅎㅎ

mnonm replied:
네,  강의 때 실습한 내용입니다 :)

5minho commented on src/test/java/qna/domain/DeleteHistoryTest.java:
```diff
@@ -0,0 +1,62 @@
+import org.junit.jupiter.api.Test;
+import qna.CannotDeleteException;
+
+public class DeleteHistoryTest {
```
테스트 이름들만 봐도 요구사항이 잘 파악되어서 좋네요 👍
`질문을 삭제 한 시각과 삭제히스토리가 생성된 시각이 같아야 한다.` 라는 테스트도 짜보셨으면 좋겠어요!

mnonm replied:
추가했습니다!

5minho reviewed (changes requested):
안녕하세요 성빈님!
3단계 미션하느라 고생 많으셨습니다.
값객체도 적극적으로 써주시고 테스트도 꼼꼼하게 잘짜주셨네요!!
조금 더 고민해보면 좋을 만한 것들 코멘트로 남겼는데 확인부탁드려요!

mnonm committed:
fix(Password): 비밀번호 유효성 검사 추가

- 널 또는 공백일 경우 예외 발생
- 테스트 추가

mnonm committed:
fix(Answers): CascadeType 수정

- PERSIST, MERGE 만 포함

mnonm committed:
fix(Answer, Question): 연관 관계 편의메소드 접근 제한자 수정

- public -> default 변경
- Answer 생성자에서 편의 메소드 호출

mnonm committed:
refactor: 접근 제한자 수정 및 미사용 함수 제거

- 미사용 setter 제거
- setter 접근제한자 수정

mnonm committed:
refactor: SavingDeleteHistory.java 제거

- 삭제 기록을 남기기 위한 인터페이스 제거, 삭제 기록 데이터가 외부에 쉽게 노출됨
- 삭제기록 생성 함수의 접근 제한자를 private으로 변경

mnonm committed:
fix(Answer, Question): 질문 삭제시 질문, 답변의 삭제 시각이 모두 동일하도록 수정

- 삭제 기록의 생성 시각도 동일하도록 수정

5minho commented on src/test/java/subway/StationTest.java:
```diff
@@ -0,0 +1,51 @@
+        final Station station2 = stations.findByName("몽촌토성역");
+        assertThat(station2).isNotNull();
+    }
+}
```
파일 끝은 새로운 라인으로 끝나게 해주는게 좋을거 같아요~
https://velog.io/@doondoony/posix-eol

5minho reviewed (approved):
안녕하세요 성빈님!
피드백 잘 반영해주셨고요
JPA 미션 진행하시느라 고생하셨어요!
마지막 미션까지 화이팅입니다!!
